src/java.base/share/classes/jdk/internal/foreign/abi/VMStorage.java

```diff
@@ -27,21 +27,35 @@
 import java.util.Objects;
 
 public class VMStorage {
-    private final int type;
+    private final byte type;
+    private final short segmentMaskOrSize;
```

Comment:
We have a couple of fields here which do double duty - e.g. depending on whether the storage models a stack argument or a record, we end up with different meanings. One way is to use (as you did) names like `segmentMaskOrSize`.

But perhaps a better (?) way would be to make VMStorage an abstract class/interface and then have two implementations, for stack slots and registers. Although I realize this might impact the VM code in a bad way (as I think this class is mirrored there in some way).

Reply:
That's ok, I suspected something similar was at play. In that case I suggest a uniform naming scheme, plus perhaps some documents explaining the semantics of the fields in the various cases.

Reply:
I can add some comments to the fields. What do you mean by 'uniform naming scheme'?

Reply:
> I can add some comments to the fields. What do you mean by 'uniform naming scheme'?

Sorry, I meant that one field is called `segmentMaskOrSize` but the other is just called `index`. Using `or` on both might be a good idea.

Reply:
Ah, ok. That sounds good.